psydac/feec/derivatives.py

```diff
@@ -38,39 +40,358 @@ def block_tostencil(M):
+            out = self._codomain.zeros()
+        
+        assert isinstance(out, StencilVector)
+        assert out.space is self._codomain
+
+        # apply the differentiation and return the result
+        self._do_diff(v, out)
+
+        return out
+    
+    def tokronstencil(self):
+        """
+        Converts this KroneckerDerivativeOperator into a KroneckerStencilMatrix.
+
+        Returns
+        -------
+        out : KroneckerStencilMatrix
+            The resulting KroneckerStencilMatrix.
+        """
+        # build derivative stencil matrix (don't care for transposition here)
+        # hence, use spaceV and spaceW instead of domain, codomain
+        periodic_d = self._spaceV.periods[self._diffdir]
+        p_d = self._spaceV.pads[self._diffdir]
+        n_d = self._spaceV.npts[self._diffdir]
+        m_d = self._spaceW.npts[self._diffdir]
+
+        V1_d = StencilVectorSpace([n_d], [p_d], [periodic_d])
+        V2_d = StencilVectorSpace([m_d], [p_d], [periodic_d])
+        M  = StencilMatrix(V1_d, V2_d)
+
+        # handle sign already here for now...
+        sign = -1. if self._negative else 1.
+        M._data[p_d:p_d+m_d, p_d]   = -1. * sign
+        M._data[p_d:p_d+m_d, p_d+1] =  1. * sign
+        
+        # now transpose, if needed
+        if self._transposed:
+            M = M.T
+
+        # identity matrices
+        def make_id(i):
+            n_i = self._domain.npts[i]
+            p_i = self._domain.pads[i]
+            periodic_i = self._domain.periods[i]
+            return IdentityStencilMatrix(StencilVectorSpace([n_i], [p_i], [periodic_i]))
+
+        # combine to Kronecker matrix
+        mats = [M if i == self._diffdir else make_id(i) for i in range(self._domain.ndim)]
+        return KroneckerStencilMatrix(self._domain, self._codomain, *mats)
+    
+    def transpose(self):
+        """
+        Transposes this operator. Creates and returns a new object.
+
+        Returns
+        -------
+        out : DirectionalDerivativeOperator
+            The transposed operator.
+        """
+        return DirectionalDerivativeOperator(self._spaceV, self._spaceW,
+                self._diffdir, negative=self._negative, transposed=not self._transposed)
+
+    @property
+    def T(self):
+        """
+        Short-hand for transposing this operator. Creates and returns a new object.
+        """
+        return self.transpose()
+    
+    def __neg__(self):
+        """
+        Negates this operator. Creates and returns a new object.
+        """
+        return DirectionalDerivativeOperator(self._spaceV, self._spaceW,
+                self._diffdir, negative=not self._negative, transposed=self._transposed)
+    
+    def toarray(self):
+        """
+        Transforms this operator into a dense matrix.
+        Includes padding in both domain and codomain.
+
+        Returns
+        -------
+        out : ndarray
+            The resulting matrix.
+        """
+        return self.tosparse().todense()
```

Comment:
Numpy's `Matrix` class is no longer recommended even for linear algebra (see [here](https://numpy.org/devdocs/reference/generated/numpy.matrix.html#numpy-matrix)), therefore you should use the `toarray` method instead.